riscv/runtime/src/coprocessors.rs

```diff
@@ -0,0 +1,10 @@
+// This is a dummy implementation of Poseidon hash,
+// which will be replaced with a call to the poseidon
+// coporocessor during compilation.
+// The function itself will be removed by the compiler
+// during the reachability analysis.
+#[no_mangle]
+#[inline(never)]
+pub fn poseidon_hash(_data: &[u8], _size: usize) -> [u8; 32] {
+    return [0; 32];
+}
```

Comment:
This function's signature is inspired from the what the keccack256 one would be, but poseidon seems way more generic than that. I would appreciate some input here.

Reply:
So in #395 (not merged yet, but working already), I implemented the basic Poseidon permutation (basically the $f$ [here](https://affine.group/images/2020/sponge.png). This lets you hash 2 field elements to one. There is an additional input (the capacity) which should be set to a constant, but it can be used for domain separation, so I think it could be exposed as well!

This already lets you build Merkle trees, which is a big use-case. The full Sponge construction is not implemented yet.

About going from bytes to field elements: How many bytes we can pack into one field element depends on the field. But I think it should be safe to assume that we can pack at least 32 bits (i.e., one RISC-V word).

Reply:
(It's working on the `poseidon2` branch, but only because it includes the changes from #393, so it wouldn't work yet on your branch).

Reply:
Thank you. @leonardoalt and I talked about this yesterday and we agreed that I will implement the sponge in rust for the time being, so that I can use your 2 -> 1 coprocesso.

Reply:
Cool! I'm not sure how it would work exactly though. In the case of BN254, the Poseidon permutation maps 3 field elements (rate `r1` & `r2`, capacity `c`) to 3 field elements.

In my understanding, the full Sponge construction should be:
```
r1, r2, c = 0, 0, 0
for m1, m2 in padded_message_blocks:
  r1, r2, c = poseidon_permutation(r1 + m1, r2 + m2, c)
return r1
```

In that case, you'd need all 3 inputs and outputs, right? (Also, in theory, you'd have to implement field addition in Rust...) I guess with 2 inputs and 1 output you could implement the Merkle–Damgård construction.

Anyway, all this is to say, let me know if you more inputs / outputs :)